src/moi_oracle.jl

```diff
@@ -53,6 +53,302 @@ function compute_extreme_point(
+                            end
+                        end
+                    end
+                    MOI.add_constraint(lmo2.o, func, MOI.EqualTo(set.lower))
+                end
+            elseif S <: MOI.LessThan
+                if set.upper ≈ val
+                    MOI.delete(lmo2.o, c_idx)
+                    if F <: MOI.VariableIndex
+                        check_cidx = MOI.ConstraintIndex{F,MOI.GreaterThan{Float64}}(c_idx.value)
+                        if MOI.is_valid(lmo2.o, check_cidx)
+                            MOI.delete(lmo2.o, check_cidx)
+                        end
+                    else
+                        func_dict = Dict(field => getfield(func, field) for field in fieldnames(typeof(func)))
+                        const_list_greater = MOI.get(lmo2.o, MOI.ListOfConstraintIndices{F,MOI.GreaterThan{Float64}}())
+                        for c_idx_greater in const_list_greater
+                            func_greater = MOI.get(lmo2.o, MOI.ConstraintFunction(), c_idx_greater)
+                            func_greater_dict = Dict(field => getfield(func_greater, field) for field in fieldnames(typeof(func_greater)))
+                            if func_greater_dict == func_dict
+                                MOI.delete(lmo2.o, c_idx_greater)
+                                break
+                            end
+                        end
+                    end
+                    MOI.add_constraint(lmo2.o, func, MOI.EqualTo(set.upper))
+                end
+            elseif S <: MOI.Interval
+                if set.upper ≈ val
+                    MOI.delete(lmo2.o, c_idx)
+                    MOI.add_constraint(lmo2.o, func, MOI.EqualTo(set.upper))
+                elseif set.lower ≈ val
+                    MOI.delete(lmo2.o, c_idx)
+                    MOI.add_constraint(lmo2.o, func, MOI.EqualTo(set.lower))
+                end
+            end
+        end
+    end
+    # Check if extra information needs to be updated
+    
+    if !isempty(solve_data)
+        for attribute in keys(solve_data)
+            val = MOI.get(lmo2.o, attribute)
+            solve_data[attribute] = val
+        end
+    end
+    MOI.set(lmo2.o, MOI.ObjectiveSense(), MOI.MIN_SENSE)
+    MOI.optimize!(lmo2.o)
+    a = MOI.get(lmo2.o, MOI.VariablePrimal(), variables)
+    MOI.empty!(lmo2.o)
+    return a
+end
+
+
+function dicg_maximum_step(lmo::MathOptLMO{OT}, direction, x; exactness=1000, atol=1e-16) where {OT}
+    gamma_max = 0.0
+    gamma = 1.0
+    precheck, on_lowerbound_idx_value,  on_upperbound_idx_value = is_constraints_feasible(lmo, x; atol)
+    if !precheck
+        error("x is not a fesible point!")
+    end
+
+    for (idx,value) in on_lowerbound_idx_value
+        if direction[idx] > value
+            return gamma_max
+        end
+    end
+
+    for (idx,value) in on_upperbound_idx_value
+        if direction[idx] < value
+            return gamma_max
+        end
+    end
+
+    while(exactness != 0)
```

Comment:
I don't understand what's the logic with the exactness counter?

Reply:
In this function, i use the bisection idea.

We start with gamma = 1.0 and we check if the resulting point satisfies the constraints, if so, return gamma. Otherwise, we shrink the gamma to 0.5 and cheak feasibility, if satisfies, gamma goes to 0.75, otherwise gamma goes to 0.25.

We repeat this process till we run out of the exactness counter.

But this process is impratical bc it requires a lot of time to check feasibility and also jump around a lot to get high precision of gamma_max.

In another dicg_maximum_step, we only check every constraint once and get the corresponding upper_bound of gamma, and we pick the samllest one, which is much faster and pratical.